tuf/api/metadata.py

```diff
@@ -154,96 +122,166 @@ def verify(self) -> bool:
-    def read_from_json(cls, filename: str) -> Metadata:
-        md = Metadata.read_from_json(filename)
-        snapshot = cls(md.consistent_snapshot, md.expiration, md.keyring, md.version)
-        meta = md._signed['meta']
-        for target_role in meta:
-            version = meta[target_role]['version']
-            length = meta[target_role].get('length')
-            hashes = meta[target_role].get('hashes')
-            snapshot.targets_fileinfo[target_role] = tuf.formats.make_metadata_fileinfo(version, length, hashes)
-        tuf.formats.SNAPSHOT_SCHEMA.check_match(snapshot.signed)
-        snapshot._signatures = md._signatures
-        return snapshot
+class Snapshot(Signed):
+    def __init__(self, meta: JsonDict = None, **kwargs) -> None:
+        # TODO: How much init magic do we want?
+        # TODO: Is there merit in creating classes for dict fields?
+        super().__init__(**kwargs)
+        self.meta = meta
 
-    @property
-    def signed(self) -> JsonDict:
-        return tuf.formats.build_dict_conforming_to_schema(
-            tuf.formats.SNAPSHOT_SCHEMA, version=self.__version,
-            expires=self.expires, meta=self.targets_fileinfo)
+    def as_dict(self) -> JsonDict:
+        json_dict = super().as_dict()
+        json_dict.update({
+            'meta': self.meta
+        })
+        return json_dict
 
     # Add or update metadata about the targets metadata.
-    def update(self, rolename: str, version: int, length: Optional[int] = None, hashes: Optional[JsonDict] = None) -> None:
-        self.targets_fileinfo[f'{rolename}.json'] = tuf.formats.make_metadata_fileinfo(version, length, hashes)
-
-class Targets(Metadata):
-    def __init__(self, consistent_snapshot: bool = True, expiration: datetime = datetime.today(), keyring: KeyRing = None, version: int = 1) -> None:
-        super().__init__(consistent_snapshot, expiration, keyring, version)
-        self.targets = {}
-        self.delegations = {}
-
-    @classmethod
-    def read_from_json(cls, filename: str) -> Metadata:
-        targets = Metadata.read_from_json(filename)
-        targets.targets = self.__signed['targets']
-        targets.delegations = self.__signed.get('delegations', {})
-        tuf.formats.TARGETS_SCHEMA.check_match(targets.signed)
-        targets._signatures = md._signatures
-        return targets
-
-    @property
-    def signed(self) -> JsonDict:
-        return tuf.formats.build_dict_conforming_to_schema(
-            tuf.formats.TARGETS_SCHEMA,
-            version=self.__version,
-            expires=self.expires,
-            targets=self.targets,
-            delegations=self.delegations)
+    def update(
+            self, rolename: str, version: int, length: Optional[int] = None,
+            hashes: Optional[JsonDict] = None) -> None:
+        metadata_fn = f'{rolename}.json'
+
+        self.meta[metadata_fn] = {'version': version}
+        if length is not None:
+            self.meta[metadata_fn]['length'] = length
```

Comment:
if it is `None`, do we need to delete an existing value?

Reply:
Good question. Is this something you and @MVrachev discussed on #1031? What do you think should happen? I understood passing `length=None` as, "I don't want length on this".  But it could also mean, "I don't want to re-assign length, use the old one."

Reply:
We did not. It seemed more apparent here because my imagined workflows for the API are different. This API feels more focused on loading and updating existing metadata, such that we'll often start from a populated class structure and update it. Whereas repository_tool feels like it's more about generating new metadata.

I think passing `length=None` should mean "I don't want a length on this", but we should think about the (unlikely?) case that i.e. a timestamp with a length is loaded, and the next version doesn't want a length.

Reply:
Not sure I fully understand. We do override the existing *fileinfo* in L260. So any previously loaded `version`, `length` and `hashes` will be discarded at any rate.

Reply:
🤦 if we're not updating the existing values what I wrote makes no sense.

Reply:
No worries! I thought you meant that someone wanted to update version only and keep the existing hashes and length. Right now this would require to do something like:
```python
obj.update(version=version, length=obj.length, hashes=obj.hashes)
```
But I think that's okay.